### PR TITLE
Add ability to edit links in rich text editor

### PR DESCRIPTION
**Pull request by loicteixeira**
See corresponding issue #444.

 - [X] If a link already exists, open the link chooser panel on the relevant tab, otherwise default to page tab.
 - [X] Pre-populate link text and url when available
 - [X] Add a remove link button (with new broken-chain icon)
 - [X] Add undo/redo support for remove link
 - ~~Add undo/redo support for add/edit link~~
 - [X] Always show the link text field for external/email tabs since the text is now pre-populated?
 - ~~On the page tab, highlight the currently selected page?~~
 - [X] Add tooltip on link over in the rich text editor?

### CODE REVIEW

**Comment by loicteixeira** [comment]
Some points on the checklist are currently written as questions. I would appreciate feedback on them.

##### Always show the link text field for external/email tabs since the text is now pre-populated?
At the moment, if your selection has some text in it, it hides the `text` field so you only had to fill the `url` one. Since it now pre-populate the fields, I think it would be great not to hide the `text` field, allowing you to update the link text on the fly.

In addition, it would simplify the python side of things which has to conditionally use different forms.

##### On the page tab, highlight the currently selected page?
With this PR changes, it will open the page chooser on the currently selected pages but there is no highlighting happening so it's difficult to know which one is already selected.

##### Remove link button icon
I used the `icon-bin` but there might be a better suited icon for this? I haven't looked at all the icons available in the `wagtail` font.

**Comment by SalahAdDin** [comment]
:+1:

**Comment by loicteixeira** [comment]
Adding `undo/redo` support for the `remove link` action was easy, but I can't get it to work for the `add/edit link` action.

For the action to register to the `undo/redo stack`, you need to use `document.execCommand`. `createLink` works but lack the support for `data` attributes which we need (you can only add an url). `insertHTML` can do that but isn't supported by IE. Any javascript dev with an idea on how to overcome this?

**Comment by JoshBarr** [comment]
> Always show the link text field for external/email tabs since the text is now pre-populated?

Yes, let's do this. Less logic involved :) 

> On the page tab, highlight the currently selected page?
This is a bit harder, without passing some additional properties to the chooser. I think for starters, having the link chooser start at the correct part of the tree is fine. 

> Add tooltip on link over in the rich text editor?
Great idea. Google docs has solved this problem really well. It's sometimes handy to quickly see (and follow) the link. 

> Remove link button icon
How about the [Broken Chain](http://fortawesome.github.io/Font-Awesome/icon/chain-broken/)?

**Comment by loicteixeira** [comment]
##### Update
- The `remove link` button now uses the `broken-chain` icon:
screen shot 2016-03-31 at 1 44 21 pm

- Hovering over a link shows a tooltip:
screen shot 2016-03-31 at 1 49 49 pm

- The external and email link chooser always show the `link text` field now, which will be pre-populated if there is a selection.

- The link chooser used to require the link text while the email link chooser would fallback to using the email as link text. The link chooser has the same behaviour now.

##### Next
- Working on the failing tests.

**Comment by loicteixeira** [comment]
Tests pass locally and will hopefully do the same on travis. I've removed the WIP tag from the PR name.

It's still missing:
 -  `undo/redo for add/edit link` but it's too much work to make it work properly with `document.execCommand` with all the browsers
 - `highlight currently selected internal page on page chooser` which is probably quite a bit of work as well. The fact that you can hover a link to know where it points should help anyway.

Let me know if there is any more work that should be done on this.

**Comment by loicteixeira** [comment]
I can see the branch has merge conflict now. Do you want me to merge master in the branch? Is there any rebase/not rebase policy?

**Comment by JoshBarr** [comment]
Hey Loïc, 

From your branch, rebasing is definitely the preferred option to keep the history free of merge commits:

```
git remote add upstream git@github.com:torchbox/wagtail.git
git fetch upstream
git rebase upstream/master
```

You might also want to consider [squashing](https://git-scm.com/book/en/v2/Git-Tools-Rewriting-History) your `fix` commits to tidy up the history a little. Once you've rebased the master into your branch, you can run an interactive rebase, and change the `pick`s to `fixup`s for any small touchup commits. This helps the whole project history look a bit cleaner. The guide linked above has some more info.

**Comment by JoshBarr** [comment]
Hey @gasman @kaedroho, since this fixes a long-standing bug in the text editor, could we release it in 1.4.X rather than waiting for 1.5? It's an extremely common client complaint for us, and we essentially need this in production as soon as possible.

I'm happy to take care of the reviewing + merging if you like.

**Comment by gasman** [comment]
@JoshBarr No, sorry. The whole point of 'patch' releases is to keep them comparatively stable by not introducing new functionality - it defeats the object of them if we start using them to mean "the next release, but faster"...

If you need to put a feature into production before it's made it into an official release, there's always the option of packaging up a custom Wagtail build - and that way, it serves as a kind of 'advance release candidate' so that the feature can receive some extra testing before it gets released to the wider world.

**Comment by loicteixeira** [comment]
Rebased master and squashed work into 2 commits.

**Comment by JoshBarr** [comment]
Loīc, I've just cleaned up the commits here – there was an issue with the rebasing that somehow collected 97 other commits 😱 .

File: wagtail/wagtailadmin/static_src/wagtailadmin/js/blocks/rich-text.js [hunk 1]
```diff
@@ -0,0 +1,13 @@
...
+            animation: false,
+            title: function() {
+                return $(this).attr('href');
+            },
+            trigger: 'hover',
+            placement: 'bottom',
+            selector: 'a'
+        });
+    };
+})(jQuery);
```
Comment (JoshBarr) [inline wagtail/wagtailadmin/static_src/wagtailadmin/js/blocks/rich-text.js]: Missing newline at end of file

File: wagtail/wagtailadmin/static_src/wagtailadmin/js/vendor/bootstrap-tooltip.js [hunk 2]
```diff
@@ -0,0 +1,386 @@
...
+
+  // TOOLTIP NO CONFLICT
+  // ===================
+
+  $.fn.tooltip.noConflict = function () {
+    $.fn.tooltip = old
+    return this
+  }
+
+}(window.jQuery);
```
Comment (JoshBarr) [inline wagtail/wagtailadmin/static_src/wagtailadmin/js/vendor/bootstrap-tooltip.js]: I know this is a library, but also missing newline at end of file :)

**Comment by loicteixeira** [comment]
Ah.. Sorry about that and thanks for fixing it.

Regarding the missing new line at the end of the file, I've pushed some changes so there is a new line, but GitHub seems to not display it 😯

**Comment by JoshBarr** [comment]
@timheap, got 5 mins to give this a look over? If it receives your blessing I shall merge + close.

**Comment by gasman** [comment]
@JoshBarr Hold on - I'd like to check this one too :-) Will hopefully have chance to do so tomorrow.

**Comment by loicteixeira** [comment]
Travis build failed because of an Elastic Search timeout. Can the build be retried?

**Comment by gasman** [comment]
Done

**Comment by gasman** [comment]
Merged in ac5aa73de97ac9037bb9882372ec10d9256cb154 + parents - thanks!

**Pull request closed without merging**